hal/armv8m/nrf/91/nrf91.c

```diff
@@ -0,0 +1,224 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * nRF91 basic peripherals control functions
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "nrf91.h"
+
+#include "../../../cpu.h"
+#include "../../armv8m.h"
+#include "../../../../include/errno.h"
+
+
+static struct {
+	volatile u32 *scb;
+	volatile u32 *power;
+	volatile u32 *clock;
+	volatile u32 *gpio;
+	u32 cpuclk;
+	spinlock_t pltctlSp;
+} nrf91_common;
+
+
+/* clang-format off */
+enum { power_tasks_constlat = 30, power_tasks_lowpwr, power_inten = 192, power_intenset, power_intenclr, power_status = 272};
+
+
+enum { clock_tasks_hfclkstart, clock_inten = 192, clock_intenset, clock_intenclr, clock_hfclkrun = 258, clock_hfclkstat };
+
+
+enum { gpio_out = 1, gpio_outset, gpio_outclr, gpio_in, gpio_dir, gpio_dirsetout, gpio_dirsetin, gpio_cnf = 128 };
+
+
+enum { syst_csr = 4, syst_rvr, syst_cvr, syst_calib };
+
+
+enum { fpu_cpacr = 34, fpu_fpccr = 141, fpu_fpcar, fpu_fpdscr };
+/* clang-format on */
+
+
+/* platformctl syscall */
+
+
+/* TODO: add platformctl implementation */
+int hal_platformctl(void *ptr)
+{
+	return 0;
+}
+
+
+void _nrf91_platformInit(void)
+{
+	hal_spinlockCreate(&nrf91_common.pltctlSp, "pltctl");
+}
+/* SysTick */
+
+
+int _nrf91_systickInit(u32 interval)
+{
+	u64 load = ((u64)interval * nrf91_common.cpuclk) / 1000000;
+	if (load > 0x00ffffff) {
+		return -EINVAL;
+	}
+
+	*(nrf91_common.scb + syst_rvr) = (u32)load;
+	*(nrf91_common.scb + syst_cvr) = 0u;
+
+	/* Enable systick */
+	*(nrf91_common.scb + syst_csr) |= 0x7u;
+
+	return EOK;
+}
+
+
+/* GPIO */
+
+
+int _nrf91_gpioConfig(u8 pin, u8 dir, u8 pull)
+{
+	if (pin > 31) {
+		return -1;
+	}
+
+	if (dir == gpio_output) {
+		*(nrf91_common.gpio + gpio_dirsetout) = (1u << pin);
+		while ((*(nrf91_common.gpio + gpio_dir) & (1u << pin)) != (1u << pin)) {
+		}
```

Comment:
Wouldn't memory barier (`dmb` instruction) suffice? I don't really know, but I suspect so. Applies to other instances.

Reply:
If no, then maybe consider using static inline function to perform write, read back and compare operation.

Reply:
In this case `dmb` is sufficient, but when checking events I'd rather leave it as it was:
```C
		while (*(console_common.base + uarte_events_txstarted) != 1u) {
		}
```

Reply:
Yeah, I was referring only to `x = a;` `while (x != a);`